### PR TITLE
Refactor WindowFunction process rows in batched partition

### PR DESCRIPTION
**Pull request by xiangfu0**
Refactor WindowAggregateOperator from row based processing model to batch processing model.
1. Partition the data to containers first
2. For each partition, generate the window agg results.

Motivation: following how presto models the WindowFunction with one `WindowFunction` interface, then 3 types:
1. AggregateWindowFunction: this is one function try to handle all the existing aggregation functions 
2. Abstract `RankingWindowFunction`, this is for all the ranking functions. Here are all the implements: `CumulativeDistributionFunction`, `CustomRank`, `DenseRankFunction`, `NTileFunction`, `PercentRankFunction`, `RankFunction`, `RowNumberFunction`
3. Abstract `ValueWindowFunction`, this is for all the row position based functions.  Here are all the implements: `FirstValueFunction`, `LagFunction`, `LastValueFunction`, `LeadFunction`, `NthValueFunction`

### CODE REVIEW

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/window/WindowFunction.java [hunk 1]
```diff
@@ -18,14 +18,75 @@
...
+import org.apache.pinot.query.runtime.operator.window.value.ValueWindowFunction;
 
 
-public interface WindowFunction {
+/**
+ * This class provides the basic structure for window functions. It provides the batch row processing API:
+ * processRows(List<Object[]> rows) which processes a batch of rows at a time.
+ *
+ */
+public abstract class WindowFunction extends AggregationUtils.Accumulator {
```
Comment (Jackie-Jiang) [inline pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/window/WindowFunction.java:41]: Suggest adding a `WindowFunctionFactory` if you want to make it pluggable, maybe following the same way of how `TransformFunction` is handle

Comment (Jackie-Jiang) [reply pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/window/WindowFunction.java:41]: Also consider making this a separate PR?

**Review by Jackie-Jiang — COMMENTED** [review]
Can you add some more comments in the PR describing about the different types of window functions extracted? What is their difference?

**Comment by xiangfu0** [comment]
> Can you add some more comments in the PR describing about the different types of window functions extracted? What is their difference?

Following how presto models the WindowFunction with one `WindowFunction` interface, then 3 types:
1. AggregateWindowFunction: this is one function try to handle all the existing aggregation functions 
2. Abstract `RankingWindowFunction`, this is for all the ranking functions. Here are all the implements: `CumulativeDistributionFunction`, `CustomRank`, `DenseRankFunction`, `NTileFunction`, `PercentRankFunction`, `RankFunction`, `RowNumberFunction`
3. Abstract `ValueWindowFunction`, this is for all the row position based functions.  Here are all the implements: `FirstValueFunction`, `LagFunction`, `LastValueFunction`, `LeadFunction`, `NthValueFunction`

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/window/WindowFunction.java (hunk 1 above)
Comment (xiangfu0) [reply pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/window/WindowFunction.java:41]: Added WindowFunctionFactory, will make a separated pr for pluggable stuffs

**Review by Jackie-Jiang — APPROVED** [review]

**Pull request merged**